HeterogeneousCore/CUDACore/interface/CUDAScopedContext.h

```diff
@@ -0,0 +1,109 @@
+#ifndef HeterogeneousCore_CUDACore_CUDAScopedContext_h
+#define HeterogeneousCore_CUDACore_CUDAScopedContext_h
+
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/EDPutToken.h"
+#include "CUDADataFormats/Common/interface/CUDAProduct.h"
+#include "HeterogeneousCore/CUDACore/interface/CUDAContextToken.h"
+
+#include <cuda/api_wrappers.h>
+
+#include <optional>
+
+namespace cudatest {
+  class TestCUDAScopedContext;
+}
+
+/**
+ * The aim of this class is to do necessary per-event "initialization":
+ * - setting the current device
+ * - calling edm::WaitingTaskWithArenaHolder::doneWaiting() when necessary
+ * - synchronizing between CUDA streams if necessary
+ * and enforce that those get done in a proper way in RAII fashion.
+ */
+class CUDAScopedContext {
+public:
+  explicit CUDAScopedContext(edm::StreamID streamID);
+
+  explicit CUDAScopedContext(CUDAContextToken&& token):
+    currentDevice_(token.device()),
+    setDeviceForThisScope_(currentDevice_),
+    stream_(std::move(token.streamPtr()))
+  {}
+
+  template<typename T>
+  explicit CUDAScopedContext(const CUDAProduct<T>& data):
+    currentDevice_(data.device()),
+    setDeviceForThisScope_(currentDevice_),
+    stream_(data.streamPtr())
+  {}
+
+  explicit CUDAScopedContext(edm::StreamID streamID, edm::WaitingTaskWithArenaHolder waitingTaskHolder):
+    CUDAScopedContext(streamID)
+  {
+    waitingTaskHolder_ = std::move(waitingTaskHolder);
```

Comment:
out of curiosity, why not
```c++
explicit CUDAScopedContext(edm::StreamID streamID, edm::WaitingTaskWithArenaHolder waitingTaskHolder):
    CUDAScopedContext(streamID),
    waitingTaskHolder_(std::move(waitingTaskHolder))
{}
```
?

Reply:
https://en.cppreference.com/w/cpp/language/initializer_list

> **Delegating constructor**
>
> If the name of the class itself appears as class-or-identifier in the member initializer list, then the list must consist of that one member initializer only

Reply:
Thanks.